homeassistant/components/network/__init__.py

```diff
@@ -0,0 +1,88 @@
+"""The Network Configuration integration."""
+from __future__ import annotations
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.core import HomeAssistant
+from homeassistant.loader import bind_hass
+
+from .const import (
+    ATTR_ADAPTERS,
+    ATTR_CONFIGURED_ADAPTERS,
+    DOMAIN,
+    NETWORK_CONFIG_SCHEMA,
+)
+from .models import Adapter
+from .network import Network
+
+ZEROCONF_DOMAIN = "zeroconf"  # cannot import from zeroconf due to circular dep
+_LOGGER = logging.getLogger(__name__)
+
+
+@bind_hass
+async def async_get_adapters(hass: HomeAssistant) -> list[Adapter]:
+    """Get the network adapter configuration."""
+    network: Network = hass.data[DOMAIN]
+    return network.adapters
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up network for Home Assistant."""
+
+    hass.data[DOMAIN] = network = Network(hass)
+    await network.async_setup()
+    await network.async_migrate_from_zeroconf(config.get(ZEROCONF_DOMAIN, {}))
+    network.async_configure()
+
+    _LOGGER.debug("Adapters: %s", network.adapters)
+
+    websocket_api.async_register_command(hass, websocket_network_adapters)
+    websocket_api.async_register_command(hass, websocket_network_adapters_configure)
+
+    return True
+
+
+@websocket_api.require_admin  # type: ignore[arg-type]
+@websocket_api.async_response
+@websocket_api.websocket_command({vol.Required("type"): "network"})
```

Comment:
It actually should be other way round. `async_response` should be applied first.
```suggestion
@websocket_api.require_admin
@websocket_api.websocket_command({vol.Required("type"): "network"})
@websocket_api.async_response
```

Reply:
The `websocket_command` decorator just stores some data on the function it wraps. All other decorators should use `@functools.wraps` which copies that info. 

The only thing that matters is that `@require_admin` is above `@async_response` because it means we check admin rights in a callback before we schedule the async command handler.

Reply:
To match types `@async_response` should be before other `websocket_api` decorators. Because it takes async and returns sync function. Other decorators don't modify function signature. That means `@websocket_command` should be also before `@async_response`.

Reply:
Adjusted in 770ce44a7d